README.md
# government domain data

Data for the prototype [government-domain-data register](http://government-domain-data.openregister.org),
a list of Government domains.


# Data

## [government-domain](data/government-domain-data.tsv)

http://government-domain.discovery.openregister.org


# Lists

A list is published annually on GOV.UK
[here](https://www.gov.uk/government/publications/list-of-gov-uk-domain-names).
The data is originally supplied by JISC, who have provided the data directly to
as a spreadsheet (not included in this repo), which we have converted to a [tsv
file](lists/jisc/list.tsv).

## [GOV.UK](lists/govuk)

[GOV.UK Government
domains](https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/557592/List_of_gov.uk_domains_as_of_1_Oct_2016.csv)

## [Rapid7](lists/rapid7)

[Rapid7's](lists/rapid7) [Rapid7 Project Sonar
(deprecated)](https://scans.io/study/sonar.fdns), via
[blog](https://shkspr.mobi/blog/2015/11/a-complete-list-of-every-uk-government-domain-name/).
[New](https://scans.io/study/sonar.rdns_v2) data is available.

The Sonar Project by Rapid7 gathers, among other things, 'the reverse DNS
records for all IPv4 addresses'.  That is effectively a list of domain names on
the internet, which can be filtered for ones ending '.gov.uk'.  A blogger [did
so](https://shkspr.mobi/blog/2015/11/a-complete-list-of-every-uk-government-domain-name/)
in November 2015 and made the method and results available.


HISTORY.md
# Register updates
## Backlog1. Request a register
June 2017

GDS met with Evans Bissessar, the Government Web Domain Manager, to discuss the potential of creating a register of gov.uk domains which are reserved for use by the public sector.

Domains on gov.uk have the format of “something”.gov.uk. A register of these would complement the government-organisation, which is a list of URLs of the form gov.uk/”something”.

Evans has oversight of the registration of all new domains, which are reviewed by the Naming and Approvals Committee (NAC), chaired by Evans. The NAC also continually reviews existing domains and removes those that no longer meet the required standards. Evans publishes the list of domains annually in October.

Day to day Evans works with a team of people to maintain the data list, that are part of Jisc, a not-for-profit company that also manages the .ac.uk domain for universities. GDS met the team at Jisc to discuss how they might create and maintain the register on behalf of Evans, as they do the existing lists of domains. Any changes that Jisc makes to the domains are first approved by the NAC, which Evans chairs.
Discovery5. Review how feedback is collected
Discovery6. Review how register is updated
Discovery2. Register is accepted
Discovery3. Agree a custodian
June 2017

The RDA team made an agreement with Evans that he would be the custodian. They also agreed that Jisc were very much partners to GDS (the Data Authority) in maintaining and updating the list.
Discovery4. Agree dataset
July - August 2017

We had further conversations with Jisc about the how the register will be updated. In the immediate term, Jisc will provide monthly updates to the RDA, who will manually update the register. In the longer term, Jisc will update the register directly from their systems, via the registers API. We also agreed that the register can proceed to the beta stage before automatic updates are in place.
June 2017

The RDA team created the discovery register based on a list of domains provided by Jisc. Jisc also provided information about the organisation in government that own each domain. Using this information, the RDA linked the list of domains to several existing registers of organisations in government, such as government-organisation, local-authority-eng, local-authority-sct and internal-drainage-board.

The discovery register has the fields: government-domain; organisation (links to the government-organisation, local-authority-eng and local-authority-sct registers); start-date; end-date

We agreed the date definitions that we wanted to use:

Start date = initial registration date (the date the domain was first delegated, i.e. added to the nameservers). End date = removal date (from removal ticket) (the date the domain was removed from the nameservers)

This means that the start date won’t change if the domain owner changes from one organisation to another. If the domain expires and is then re-registered, there will be two entries (one with an end date, and one without).
Alpha7. Meet operational standards
Alpha8. Meet technical standards
Alpha9. Find duplicate lists
Beta10. Review feedback from alpha
Beta11. Remove duplicate lists
Managed by
Cabinet Office
Custodian: Evans Bissessar
Register: https://government-domain.register.gov.uk


src/jisc.R
# Convert the spreadsheet provided by JISC to tsv, omitting some columns.

# Populate 'organisation' with curies to government-organisation,
# local-authority-eng, and local-authority-sct, by:
# * exact match (889)
# * exact match after removing punctuation and stopwords (58)
# * fuzzy match confirmed manually (54 confirmed out of 485)
# * no match: 1910, but that's expected because we don't have
#   principal-local-authority (Wales) or local-authority-ni yet, and many
#   organisations aren't among even them.

library(tidyverse)
library(readxl)
library(stringi)
library(fuzzyjoin)
library(tidytext)

tsv_path <- "../data/government-domain-data.tsv"

# Function to remove stopwords and punctuation
# * of
# * for
# * the
# * and
# * (
# * )
# * &
# * ,
# * '
clean <- function(string) {
  string %>%
    tolower %>%
    stri_replace_all_fixed(c(" of ", " for ", " the ", " and ", "(", ")",
                             " & ", ", ", "'", "-"),
                           c(" ",    " ",     " ",     " ",     "",  "",
                             " ",   " ", "",   " "),
                         vectorize_all = FALSE) %>%
    stri_trim()
}

# Load the JISC-supplied spreadsheet
jisc <-
  read_excel("../lists/jisc/jisc.xlsx", skip = 3) %>%
  filter(!is.na(`Domain Name`),
         Status != "Suspended") %>% # Omit blanks due to merged cells
  select(`Domain Name`,
         `Owner`) %>%
  rename(`government-domain` = `Domain Name`,
         org_original = `Owner`) %>%
  mutate(`organisation` = clean(org_original ))

write_tsv(jisc, "../lists/jisc/jisc.tsv")

# Download the government-organisation register
gov_org <-
  read_csv("https://government-organisation.register.gov.uk/records.csv?page-size=5000") %>%
  select(`name`,
         `government-organisation`) %>%
  rename(`organisation` = `name`,
         curie = `government-organisation`) %>%
  mutate(curie = paste0("government-organisation:", curie))

# Download the various local-authority registers
la_eng <-
  read_csv("https://local-authority-eng.register.gov.uk/records.csv?page-size=5000") %>%
  select(`official-name`,
         `local-authority-eng`) %>%
  rename(`organisation` = `official-name`,
         curie = `local-authority-eng`) %>%
  mutate(curie = paste0("local-authority-eng:", curie))

la_sct <-
  read_csv("https://local-authority-sct.register.gov.uk/records.csv?page-size=5000") %>%
  select(`official-name`,
         `local-authority-sct`) %>%
  rename(`organisation` = `official-name`,
         curie = `local-authority-sct`) %>%
  mutate(curie = paste0("local-authority-sct:", curie))

# Wales and Northern Ireland aren't in yet in beta

# Combine government organisations and local authorities
organisation <-
  bind_rows(gov_org, la_eng, la_sct) %>%
  rename(org_original = organisation) %>%
  mutate(organisation = clean(org_original))

# Fuzzyjoin domain organisations to all organisations
# authorities

exact_match <- # exact match by the given organisation names
  jisc %>%
  inner_join(organisation, by = "org_original") %>%
  select(`government-domain`,
         org_original,
         curie)

jisc_remainder <- anti_join(jisc, exact_match, by = "org_original")

exact_match_stopped <- # exact match by stop-worded organisation names
  jisc_remainder %>%
  inner_join(organisation, by = "organisation") %>%
  select(`government-domain`,
         org_original = org_original.x,
         curie)

jisc_remainder <- anti_join(jisc_remainder, exact_match_stopped, by = "org_original")

fuzzy <- # fuzzy match by stop-worded organisation names
  jisc_remainder %>%
  filter(!is.na(organisation)) %>% # Can't fuzzyjoin on NA
  stringdist_left_join(organisation,
                       by = "organisation",
                       method = "jw",
                       max_dist = .18,
                       distance_col = "score") %>%
  select(-organisation.x, -organisation.y) %>%
  filter(!is.na(org_original.y)) %>% # No match met the threshold
  # Select only the best match
  group_by(org_original.x) %>%
  arrange(org_original.x, score) %>%
  slice(1) %>%
  ungroup() %>%
  arrange(org_original.y, org_original.x) # Nicer to read

write_csv(fuzzy, "fuzzy.csv")

# MANUAL STEP HERE!

# Then continue making the tsv

manual_match <- read_csv("./manual-match.csv")

matched <-
  bind_rows(exact_match,
            exact_match_stopped,
            manual_match) %>%
  select(`government-domain`,
         `organisation` = curie) %>%
  mutate(`start-date` = NA,
         `end-date` = NA)

unmatched <-
  jisc %>%
  anti_join(matched, by = "government-domain") %>%
  select(`government-domain`)

# Write the register tsv file, using write.table instead of write_tsv in order
# to quote the 'area' column.  Unfortunately doing so means that the headers are
# all quoted, so the headers are written separately.
register <-
  bind_rows(matched, unmatched) %>%
  arrange(`government-domain`) %>%
  # structure it more like https://government-organisation.alpha.openregister.org/records
  mutate(hostname = stri_replace(`government-domain`, regex = ".gov.uk$", "")) %>%
  mutate(`government-domain` = row_number() + 1000) %>%
  select(`government-domain`,
         hostname,
         organisation,
         `start-date`,
         `end-date`)

cat(paste0(paste(colnames(register), collapse = "\t"), "\n"),
    file = tsv_path)

write.table(register,
            tsv_path,
            sep = "\t",
            row.names = FALSE,
            quote = 3,
            col.names = FALSE,
            append = TRUE,
            na = "")


R/org-map.R
# Build a map from JISC organisation names to CURIEs.

library(tidyverse)
library(readxl)
library(stringr)
library(RegistersClientR)
library(here)

# The original JISC file that was mapped to CURIES
jisc_2017_06_08 <-
  read_excel(here("lists", "jisc", "jisc-2017-06-08.xlsx"), skip = 2) %>%
  rename(hostname = `Domain Name`) %>%
  select(hostname, Owner, Representing) %>%
  mutate(hostname = str_replace(hostname, "\\.gov\\.uk$", ""))

# The original register in Alpha, with owners mapped to CURIES
register_2017_07_14 <-
  read_tsv("https://raw.githubusercontent.com/openregister/government-domain-data/c1b72ed282259ae986bf3b70b1f137891dd6f089/data/government-domain-data.tsv")

# A map from owners to CURIEs, at the initial mapping
curie_2017_07_14 <-
  register_2017_07_14 %>%
  filter(!is.na(organisation)) %>%
  select(hostname, organisation) %>%
  inner_join(jisc_2017_06_08, by = "hostname")

# Check for different CURIEs mapped to the same owner
curie_2017_07_14 %>%
  distinct(Owner, organisation) %>%
  count(Owner, sort = TRUE) %>%
  filter(n > 1)

# Examine where different CURIEs were mapped to the same owner
curie_2017_07_14 %>%
  filter(Owner %in% c("Department for Communities", "HM Government Cabinet Office")) %>%
  arrange(Owner, organisation, hostname) %>%
  print(n = Inf)

# I think someone used the Representing field instead of the Owner field, when
# looking up CURIEs.  They should be corrected as follows (file in
# lists/map-correction.tsv.

map_correction <-
  read_tsv(here("lists", "map-correction.tsv")) %>%
  mutate(`start-date` = strftime(`start-date`, "%Y-%m-%d"),
         `end-date` = strftime(`end-date`, "%Y-%m-%d"))

map_correction
# # A tibble: 6 x 5
#   `government-domain`                  hostname                   organisation `start-date` `end-date`
#                 <int>                     <chr>                          <chr>        <chr>      <chr>
# 1                7448            volcomgrantsni government-organisation:OT1173   2003-02-12       <NA>
# 2                6539                     proni government-organisation:OT1173   2002-10-09       <NA>
# 3                4762                     dsdni government-organisation:OT1173   2005-04-18       <NA>
# 4                4637                    dcalni government-organisation:OT1173   2006-08-18       <NA>
# 5               15597 boundarycommissionengland  government-organisation:PB337   2014-11-05       <NA>
# 6               15644                     acoba     government-organisation:D2   2014-11-27       <NA>

# Build the CURIE map without those errors (this time include unmapped owners)
curie_map_2017_07_14 <-
  register_2017_07_14 %>%
  filter(!(hostname %in% c("dcalni", "dsdni", "proni", "volcomgrantsni", "acoba", "boundarycommissionengland"))) %>%
  inner_join(jisc_2017_06_08, by = "hostname") %>%
  distinct(Owner, organisation)

# The 2017-09-06 JISC file to be mapped to CURIES
jisc_2017_09_06 <-
  read_excel(here("lists", "jisc", "Government_Domain_Register_Report 06Sept2017.xlsx")) %>%
  rename(hostname = `Domain Name`) %>%
  select(hostname, Owner, Representing) %>%
  mutate(hostname = str_replace(hostname, "\\.gov\\.uk$", ""))

jisc_2017_09_06 %>%
  anti_join(curie_map_2017_07_14, by = "Owner") %>%
  print(n = Inf)

# New maps to be created for the 2017-09-06 data:

# Natural Resources Wales = government-organisation:OT1103
# Innovate UK = government-organisation:PB1078

curie_map_2017_09_06 <-
  bind_rows(curie_map_2017_07_14,
            tribble(                            ~Owner,                    ~organisation,
                         "Alderminster Parish Council",                               NA,
                             "Natural Resources Wales", "government-organisation:OT1103",
                        "Batchworth Community Council",                               NA,
                             "St.Agnes Parish Council",                               NA,
                         "High Halstow Parish Council",                               NA,
                             "Orgreave Parish Council",                               NA,
                              "Sancton Parish Council",                               NA,
                             "Ringmore Parish Council",                               NA,
                                "Crich Parish Council",                               NA,
                                         "Innovate UK", "government-organisation:PB1078",
                               "Milton Parish Council",                               NA,
                   "Bridgwater Without Parish Council",                               NA,
                         "Long Melford Parish Council",                               NA,
                      "Hemingford Grey Parish Council",                               NA,
            "Itchen Stoke and Ovington Parish Council",                               NA,
                              "New Mills Town Council",                               NA,
                        "Batchworth Community Council",                               NA,
                             "St Teath Parish Council",                               NA,
                              "Measham Parish Council",                               NA,
                            "West Hill Parish Council",                               NA,
              "Nythe, Eldene and Liden Parish Council",                               NA))

# The 2017-10-05 JISC file to be mapped to CURIES
jisc_2017_10_05 <-
  read_excel(here("lists", "jisc", "Government_Domain_Register_Report 5Oct2017.xlsx")) %>%
  rename(hostname = `Domain Name`) %>%
  select(hostname, Owner, Representing) %>%
  mutate(hostname = str_replace(hostname, "\\.gov\\.uk$", ""))

jisc_2017_10_05  %>%
  anti_join(curie_map_2017_09_06, by = "Owner") %>%
  print(n = Inf)

# New maps to be created for the 2017-09-06 data:

# Natural Resources Wales = government-organisation:OT1103
# Innovate UK = government-organisation:PB1078

curie_map_2017_10_05 <-
  bind_rows(curie_map_2017_07_14,
            tribble(               ~Owner, ~organisation,
        "Hampshire Fire & Rescue Service",            NA,
                "Padworth Parish Council",            NA,
       "Moss and District Parish Council",            NA,
                  "Lutton Parish Council",            NA,
                 "Holford Parish Council",            NA,
                  "Exning Parish Council",            NA,
"Stoke Lodge & The Common Parish Council",            NA,
   "Central Swindon North Parish Council",            NA,
            "West Swindon Parish Council",            NA,
               "Quethiock Parish Council",            NA,
    "Sturminster Marshall Parish Council",            NA,
        "The Small Business Commissioner",            NA,
              "Greater Manchester Police",            NA))

org_map <- curie_map_2017_10_05


src/review-match.R
# Review the matching algorithm and its success, having manually checked every
# record.

library(tidyverse)
library(googlesheets)
library(stringr)

# One-off, or to switch users)
# gs_auth(new_user = TRUE)

before <-
  read_tsv("./before-manual-match.tsv") %>%
  select(hostname, organisation)
after <-
  read_tsv("./after-manual-match.tsv") %>%
  select(hostname, organisation)
org_names <-
  read_tsv("../lists/jisc/jisc.tsv") %>%
  rename(hostname = `government-domain`,
         org_name = org_original) %>%
  select(hostname, org_name) %>%
  mutate(hostname = str_replace(hostname, "\\.gov\\.uk$", ""))

curie <- function(x) {
  str_extract(x, "^[^:]+")
}

errors <-
  before %>%
  inner_join(after, by = "hostname") %>%
  inner_join(org_names, by = "hostname") %>%
  mutate(curie.x = curie(organisation.x),
         curie.y = curie(organisation.y))
write_tsv(errors, "errors.tsv")

# None were the wrong register, good!
errors %>%
  filter(!(curie.x == curie.y))

# None were the wrong organisation.  Too cautious?
errors %>%
  filter(!(organisation.x == organisation.y))

# 999 were auto-matched or auto-assisted
errors %>%
  filter(!is.na(organisation.x))

# 1901 were not auto-matched
errors %>%
  filter(is.na(organisation.x))

# Of which, 1584 weren't even manually matched
errors %>%
  filter(is.na(organisation.x),
         is.na(organisation.y))

# So only 327 further matches were found 100% manually
errors %>%
  filter(is.na(organisation.x),
         !is.na(organisation.y))

total <- 2911
exact_match <- 889 + 58 # 58 found after stopword/punctuation/lowercase
fuzzy_attempted <- 249
fuzzy_confirmed <- 53
manual <- 327
no_match <- 1584
yes_match <- total - no_match

exact_match + fuzzy_confirmed + manual + no_match # Equals the total

manual / yes_match # 24% of all matches were found manually
exact_match / yes_match # 71% matched exactly (including stopword/puntuation/lowercase)
fuzzy_confirmed / yes_match # 4% matched by real data science + human confirmation


src/jisc-compare-lists.R
# JISC sent an updated list.  This script compares the two.

library(tidyverse)
library(readxl)

# Load the two JISC-supplied spreadsheets
jisc1 <-
  read_excel("../lists/jisc/jisc.xlsx", skip = 3) %>%
  filter(!is.na(`Domain Name`),
         Status != "Suspended") # Omit blanks due to merged cells
jisc2 <-
  read_excel("../lists/jisc/jisc-2017-06-08.xlsx", skip = 3) %>%
  filter(!is.na(`Domain Name`),
         Status != "Suspended") # Omit blanks due to merged cells

jisc2 <- read_excel("../lists/jisc/jisc.xlsx", skip = 3)
  filter(!is.na(`Domain Name`),
         Status != "Suspended") %>% # Omit blanks due to merged cells
  select(`Domain Name`,
         `Owner`) %>%
  rename(`government-domain` = `Domain Name`,
         org_original = `Owner`) %>%
  mutate(`organisation` = clean(org_original ))



R/monthly-reconciliation.R
# Script to reconcile the latest monthly list with the previous one.
# Finds records that have been:
# * created
# * deleted
# * updated

# The point of comparison is the *previouus spreadsheet*, not the *current
# register*, because JISC still uses its own names.

# This is hardcoded to the September reconciliation, because the new file has a
# different structure from the original one.



# 16:54]
# Duncan Garmonsway Reviewing your pull request for gov-domains.  We might have been at cross-purposes when we discussed the 'owner' and 'representing' column.  Neither of them should be in the register -- either we find a CURIE for the owner to put in the 'organisation' column, or we forget about it.  Does that make sense?  Do you agree?  Unlike the public-body register, which has to have at least a free-text 'name' column otherwise there's nothing left -- for gov-domain we always have the domain, and don't absolutely require anything else.


# [16:59]
# I can do my own PR if you like?


# [17:01]
# The good news is that we agree on additions/deletions :slightly_smiling_face:


# [17:12]
# Duncan Garmonsway Small number of owner changes too.
# ```               `Domain Name`                              Owner.x                              Owner.y
#                        <chr>                                <chr>                                <chr>
# 1     alderminster-pc.gov.uk   Stratford-on-Avon District Council          Alderminster Parish Council
# 2       batchworth-pc.gov.uk            Batchworth Parish Council         Batchworth Community Council
# 3                 hsl.gov.uk         Health and Safety Laboratory          Health and Safety Executive
# 4 llwybrarfordircymru.gov.uk        Countryside Council for Wales              Natural Resources Wales
# 5          pesticides.gov.uk Food and Environment Research Agency          Health and Safety Executive
# 6      walescoastpath.gov.uk        Countryside Council for Wales              Natural Resources Wales
# 7    westofengland-ca.gov.uk                 Bristol City Council Bath and North East Somerset Council```

library(tidyverse)
library(readxl)
library(RegistersClientR)
library(stringr)
library(here)

list_old <-
  read_excel((here("lists", "jisc", "Government_Domain_Register_Report 06Sept2017.xlsx"))) %>%
  mutate(`Domain Name` = str_replace(`Domain Name`, "\\.gov\\.uk$", ""))

list_new <- read_excel(file.path(here("lists", "jisc", "Government_Domain_Register_Report 5Oct2017.xlsx"))) %>%
  mutate(`Domain Name` = str_replace(`Domain Name`, "\\.gov\\.uk$", ""))

current <- read_tsv(here("data", "government-domain-data.tsv"))

# Are any current records end-dated?
filter(current, !is.na(`end-date`))
# No.  Why not?  Surely some were deleted in a previous update?

# Does the new list have start and end dates?
anyNA(list_new$`Start Date`)
anyNA(list_new$`End Date`) # No end dates.  Check with Jamie and then JISC.

# Has anything changed between list_old and current -- i.e. via custodian update
# tool?
anti_join(list_old, current, by = c("Domain Id" = "government-domain"))
# Yes, 4710 digitv is missing from current that was in list_old.  Is it in
# list_new?
filter(list_new, `Domain Id` == 4710)
filter(list_new, `Domain Name` == "digitv")
# No, so it should be added with an end-date of this month.
add1 <- data_frame(`government-domain` = 4710,
                   hostname = "digitv",
                   organisation = NA,
                   `start-date` = "2017-09-06",
                   `end-date` = "2017-10-04")

anti_join(current, list_old, by = c("government-domain" = "Domain Id"))
# Also there are several domains in current that aren't in list_old.  Digging
# through the git log, it seems there's an uncommited JISC file that added start
# dates and some more domains to the one that was committed.

# Are those domains still in list_new?
anti_join(current, list_old, by = c("government-domain" = "Domain Id")) %>%
  inner_join(list_new, by = c("government-domain" = "Domain Id"))
# Yes, so they will be left, and can be ignored.

# Do any IDs have new hostnames?
inner_join(current, list_old, by = c("government-domain" = "Domain Id")) %>%
  filter(`Domain Name` != hostname)
# No.

# Do any IDs have new owners?  We compare via the org_map dataframe, which is
# created by the R/org-map.R script.
list_old %>%
  left_join(org_map) %>%
  inner_join(filter(current, !is.na(organisation)),
             by = c("Domain Id" = "government-domain")) %>%
  filter(organisation.x != organisation.y) %>%
  select(`Domain Id`, `Domain Name`, `Owner`, organisation.x, organisation.y)
# Yes, but these will be corrected by the org_map correction.

# Has anything changed between list_new and current?
anti_join(list_new, current, by = c("Domain Id" = "government-domain"))
# Yes, there are some new entries, written to lists/new-entries.tsv
add2 <-
  anti_join(list_new, current, by = c("Domain Id" = "government-domain")) %>%
  rename(`government-domain` = `Domain Id`,
         hostname = `Domain Name`,
         `start-date` = `Start Date`) %>%
  mutate(organisation = NA,
         `start-date` = strftime(`start-date`, "%Y-%m-%d"),
         `end-date` = NA) %>%
  select(`government-domain`,
         hostname,
         organisation,
         `start-date`,
         `end-date`)

# Some entries have also gone
anti_join(current, list_new, by = c("government-domain" = "Domain Id"))
delete1 <-
  anti_join(current, list_new, by = c("government-domain" = "Domain Id")) %>%
  mutate(`start-date` = strftime(`start-date`, "%Y-%m-%d"),
         `end-date` = "2017-10-04")

# Do any IDs have new hostnames?
inner_join(current, list_new, by = c("government-domain" = "Domain Id")) %>%
  filter(`Domain Name` != hostname)
# No.

# Do any IDs have new owners?  We compare via the org_map dataframe, which is
# created by the R/org-map.R script.
list_new %>%
  left_join(org_map) %>%
  inner_join(filter(current, !is.na(organisation)),
             by = c("Domain Id" = "government-domain")) %>%
  filter(organisation.x != organisation.y) %>%
  select(`Domain Id`, `Domain Name`, `Owner`, organisation.x, organisation.y)
# Yes, but these will be corrected by the org_map correction.

bind_rows(map_correction, add1, add2, delete1) %>%
  print(n = Inf) %>%
  write_tsv(here("lists", "append-entries-2017-10-05.tsv"), na = "")
